### PR TITLE
Tuxedo Control Center

### LINKED ISSUE #132206: Packaging Request: Tuxedo Control Center
**Project description**

[Tuxedo](https://www.tuxedocomputers.com/) is a German laptop manufacturer that provides Linux-friendly laptops. Their system control is done via an app called "Tuxedo Control Center" (TCC). This open source app provides fan control settings among other things. Without this app, the Tuxedo laptops default to very noisy fan control settings. It lives on [Github](https://github.com/tuxedocomputers/tuxedo-control-center).

Besides the source, Tuxedo provides binary packages for Ubuntu and OpenSUSE.

TCC consists of:
- a set of systemd services implemented in Javascript,
- an Electron GUI application to control everything.

**Metadata**

* homepage URL: https://www.tuxedocomputers.com/en/TUXEDO-Control-Center.tuxedo
* source URL: https://github.com/tuxedocomputers/tuxedo-control-center
* license: GPL3
* platforms: Linux

### PR DESCRIPTION
**Pull request by blitz**
###### Motivation for this change

[Tuxedo](https://www.tuxedocomputers.com/) is a German laptop manufacturer that provides Linux-friendly laptops. Their system control is done via an app called "Tuxedo Control Center" (TCC). This open source app provides fan control settings among other things. Without this app, the Tuxedo laptops default to very noisy fan control settings. It lives on [Github](https://github.com/tuxedocomputers/tuxedo-control-center).

This PR:
- adds the Tuxedo Control Center application and daemon, and
- a tuxedo-control-center module to enable the daemon at boot,
- bumps tuxedo-keyboard to its latest version.

Testing done:

I've been running TCC on a Tuxedo XP14 laptop without issues. Fan control notably improves compared to running it without TCC.

This PR is based to a large degree on work done by @Vodurden.

Fixes #132206.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Mic92 — CHANGES_REQUESTED** [review]
Please integrate the node package into our `nodePackages` in `pkgs/development/node-packages` set. 12000 lines of code for a single package is too much.

**Comment by blitz** [comment]
> Please integrate the node package into our `nodePackages` in `pkgs/development/node-packages` set. 12000 lines of code for a single package is too much.

@Mic92 So in general I agree. The problem is that TCC needs its `devDependencies` to build which are not included in `nodePackages`. This results in build failure:

```
Error: Cannot find module 'node-addon-api'
Require stack:
- /nix/store/xhnnim063mxkv7y2810yr75zn7j429r5-tuxedo-control-center-1.0.14/lib/node_modules/tuxedo-control-center/[eval]
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:889:15)
    at Function.Module._load (internal/modules/cjs/loader.js:745:27)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at [eval]:1:1
    at Script.runInThisContext (vm.js:134:12)
    at Object.runInThisContext (vm.js:310:38)
    at internal/process/execution.js:81:19
    at [eval]-wrapper:6:22
    at evalScript (internal/process/execution.js:80:60) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/nix/store/xhnnim063mxkv7y2810yr75zn7j429r5-tuxedo-control-center-1.0.14/lib/node_modules/tuxedo-control-center/[eval]'
  ]
```

Other people seem to have this problem when packaging as well (see https://github.com/svanderburg/node2nix/issues/149), without any clear solution. Do you have a pointer?

**Comment by blitz** [comment]
@Mic92 I've tried to integrate TCC into `nodePackages`, but there are two roadblocks:

- TCC needs dev dependencies to be avaiable to do its build, this could probably worked around by patching the `package.json`, but it feels very hacky.
- it seems that the way `nodePackages` is composed, it doesn't respect TCC's `package-lock.json`, so you end up with incompatible libraries / untested library combinations.

So unless I'm missing something obvious packaging, TCC apart from `nodePackages` seems to be the way to go. It seems I'm also not in uncharted territory here, because there are quite a few Node/Electron packages that go the same route:

```
% find . -name "node-packages.nix" | wc -l
25
```

Don't understand me wrong, I would be happy if integrating it into `nodePackages` works, but I really don't see how. :(

**Review by blanky0230 — APPROVED** [review]

File: pkgs/os-specific/linux/tuxedo-control-center/update.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/update.sh:1]: Please use nix-shell with curl and node2nix.

File: pkgs/os-specific/linux/tuxedo-control-center/update.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+set -eu -o pipefail
+
+SCRIPTDIR="$( dirname "${BASH_SOURCE[0]}" )"
+NIXPKGS="$SCRIPTDIR"/../../../..
+node2nix="$(nix-build "$NIXPKGS" -A nodePackages.node2nix)/bin/node2nix"
+curl="$(nix-build "$NIXPKGS" -A curl)/bin/curl"
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/update.sh:8]:
```suggestion
```

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  #
+  # It also hardcodes binary path.
+  postPatch = ''
+    substituteInPlace src/common/classes/TccPaths.ts \
+      --replace "/etc/tcc" "/var/lib/tcc" \
+      --replace "/opt/tuxedo-control-center/resources/dist/tuxedo-control-center/data/service/tccd" "$out/bin/tccd"
+   '';
+
+  buildPhase = ''
+    set -x
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:76]:
```suggestion
```

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  nativeBuildInputs = [
+    desktop-file-utils # for desktop-file-validate
+  ];
+
+  buildInputs = [
+    nodejs
+    makeWrapper
+
+    # For node-gyp
+    python
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:62]:
```suggestion
    python2
```

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    mkdir -p $out/etc/dbus-1/system.d/
+    cp $out/data/dist-data/com.tuxedocomputers.tccd.conf $out/etc/dbus-1/system.d/com.tuxedocomputers.tccd.conf
+
+    # Put our icons in the right spot
+    mkdir -p $out/share/icons/hicolor/scalable/apps/
+    cp $out/data/dist-data/tuxedo-control-center_256.svg \
+       $out/share/icons/hicolor/scalable/apps/tuxedo-control-center.svg
+
+    ${desktopItem.buildCommand}
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:149]:
```suggestion
    runHook postInstall
```

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    # We need to tell npm where to find node or `node-gyp` will try to download it.
+    # This also _needs_ to be lowercase or `npm` won't detect it
+    export npm_config_nodedir=${nodejs}
+    npm run build-native   # Builds to ./build/Release/TuxedoIOAPI.node
+
+    npm run build-ng-prod
+  '';
+
+  installPhase = ''
+    mkdir -p $out
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:103]:
```suggestion
    runHook preInstall

    mkdir -p $out
```

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    ELECTRON_SKIP_BINARY_DOWNLOAD=1;
+
+    # Angular prompts for analytics, which in turn fails the build.
+    #
+    # We can disable analytics using false or empty string
+    # (See https://github.com/angular/angular-cli/blob/1a39c5202a6fe159f8d7db85a1c186176240e437/packages/angular/cli/models/analytics.ts#L506)
+    NG_CLI_ANALYTICS="false";
+  };
+
+  desktopItem = makeDesktopItem {
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:39]: Please use copyDesktopItems.

File: pkgs/os-specific/linux/tuxedo-control-center/README.md [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# How to update
+
+1. Bump the version of `tuxedo-control-center` in [packages.json](./packages.json) to the latest on [tuxedocomputers/tuxedo-control-center](https://github.com/tuxedocomputers/tuxedo-control-center)
+2. Run `generate-dependencies.sh`
+3. Set the version in [default.nix](./default.nix). (Version can be found in [tuxedocomputers/tuxedo-control-center/package.json](https://github.com/tuxedocomputers/tuxedo-control-center/blob/master/package.json))
+4. Build and test
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/README.md:6]: Maybe move this inline next to version so that reviewer also see it.

Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/README.md]: Done.

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix (hunk 7 above)
Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/default.nix]: Done.

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix (hunk 6 above)
Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/default.nix:123]: I've added pre/post hooks to build and install.

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix (hunk 4 above)
Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/default.nix]: I've changed this to use python3, which works fine as well. No reason to add more dependencies on python2.

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix (hunk 3 above)
Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/default.nix]: Done.

File: pkgs/os-specific/linux/tuxedo-control-center/update.sh (hunk 1 above)
Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/update.sh]: Done.

**Comment by blitz** [comment]
@ofborg build tuxedo-control-center

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 9]
```diff
@@ -0,0 +1,179 @@
+{ pkgs, lib, stdenv, copyDesktopItems,
+
+  python3, udev,
+
+  makeWrapper, nodejs, electron_11, fetchFromGitHub
+}:
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:6]:
```suggestion
{ pkgs, lib, stdenv, copyDesktopItems
, python3, udev
. makeWrapper, nodejs, electron_11, fetchFromGitHub
}:
```

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+let
+  baseName = "tuxedo-control-center";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:9]:
```suggestion
  pname = "tuxedo-control-center";
```

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    #
+    # We can disable analytics using false or empty string
+    # (See https://github.com/angular/angular-cli/blob/1a39c5202a6fe159f8d7db85a1c186176240e437/packages/angular/cli/models/analytics.ts#L506)
+    NG_CLI_ANALYTICS="false";
+  };
+
+in
+
+stdenv.mkDerivation rec {
+  name = "${baseName}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/tuxedo-control-center/default.nix:53]:
```suggestion
  inherit pname version;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix (hunk 11 above)
Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/default.nix]: Fixed.

File: pkgs/os-specific/linux/tuxedo-control-center/default.nix (hunk 10 above)
Comment (blitz) [reply pkgs/os-specific/linux/tuxedo-control-center/default.nix]: Fixed.

**Comment by blitz** [comment]
@SuperSandro2000 I've resolved the issues. Please have another look. :)

**Comment by blitz** [comment]
I've rebased this and updated Tuxedo Control Center to 1.1.2.

**Comment by MichaelAkvo** [comment]
Is there a way I can use this while waiting for it to get merged?

**Comment by pgronkievitz** [comment]
> Is there a way I can use this while waiting for it to get merged?
> 
> P.S thanks a lot for the work!

@blitz has made a module (unfortunately it's not a flake, but I've managed to work around this)

https://github.com/blitz/tuxedo-nixos

**Comment by blitz** [comment]
> > Is there a way I can use this while waiting for it to get merged?
> > 
> > P.S thanks a lot for the work!
> 
> @blitz has made a module (unfortunately it's not a flake, but I've managed to work around this)
> 
> https://github.com/blitz/tuxedo-nixos

I'm slowly flakifying my NixOS configs so this repo will also become a flake soon.

**Comment by MichaelAkvo** [comment]
@blitz you the real MVP.

Thanks for pointing me to the repo @pgronkievitz

**Comment by dasj19** [comment]
@MichaelAkvo if you are using the `unstable` branch of nixpkgs you can use the PR author's branch as your base for nixpkgs.
You can clone his fork and switch to the tuxedo branch then run this:
`nixos-rebuild -I nixpkgs=/path/to/your/version/of/nixpkgs` switch

As for myself I have some migration work to do on my tuxedo computer then I'll start using this and come with feedback

**Review by dasj19 — COMMENTED** [review]
There is a merge conflict and the following message
`error: Package ‘electron-11.5.0’ in /home/daniel/workspace/projects/nixpkgs/pkgs/development/tools/electron/generic.nix:25 is marked as insecure, refusing to evaluate.`

I added the electron package to the exceptions for now and started building.

Edit: build finished and got this result:
```

/bin/sh: line 1: prime-select: command not found
/bin/sh: line 1: prime-supported: command not found
(node:74258) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
(node:74258) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'on' of undefined
    at TccDBusController.onModeReapplyPendingChanged (/nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/common/classes/TccDBusController.js:206:24)
    at /nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/e-app/main.js:150:17
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:74258) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:74258) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by blitz** [comment]
@dasj19 Thanks for testing!

> There is a merge conflict 

I've rebased the PR, so this should be gone. If you actually want to use it while it's not merged, I would suggest you head over to [this repo](https://github.com/blitz/tuxedo-nixos/).

> and the following message `error: Package ‘electron-11.5.0’ in /home/daniel/workspace/projects/nixpkgs/pkgs/development/tools/electron/generic.nix:25 is marked as insecure, refusing to evaluate.`

Checkout [blitz/tuxedo-nixos#troubleshooting](https://github.com/blitz/tuxedo-nixos/#troubleshooting) and https://github.com/tuxedocomputers/tuxedo-control-center/issues/148. 

> 
> I added the electron package to the exceptions for now and started building.
> 
> Edit: build finished and got this result:
> 
> ```
> 
> /bin/sh: line 1: prime-select: command not found
> /bin/sh: line 1: prime-supported: command not found

This is harmless, unless you have a nVidia PRIME system. In this case, please say so. It would be interesting to see whether we get this to work.

> (node:74258) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
> (node:74258) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'on' of undefined
>     at TccDBusController.onModeReapplyPendingChanged (/nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/common/classes/TccDBusController.js:206:24)
>     at /nix/store/nd10h9900zy5fvzgsc282n6c5hripn1b-tuxedo-control-center-1.1.2/e-app/e-app/main.js:150:17
>     at processTicksAndRejections (internal/process/task_queues.js:97:5)
> (node:74258) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> (node:74258) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> ```

These need to be reported upstream. If there is specific broken functionality, we can try to patch it. I would welcome help for this.

**Comment by blitz** [comment]
@pgronkievitz If you want to give the flake a shot, there is a flake branch here that exports a module: https://github.com/blitz/tuxedo-nixos/tree/flake

**Comment by dasj19** [comment]
@blitz Thanks for your response and fix.

I tried the new changes and it behaves as follows:
1. When I put `tuxedo-control-center` in `environment.systemPackages` in my configuration.nix it ends with the same error you said that should be sent upstream
2. when I build with `hardware.tuxedo-control-center.enable = true;` that error does not show up but instead  I get only:
```
/bin/sh: line 1: prime-select: command not found
/bin/sh: line 1: prime-supported: command not found
(node:37258) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
```
And a blank Tuxedo Control Center shows up.

The same happens when I use https://github.com/blitz/tuxedo-nixos/#usage

My hardware us this https://www.tuxedocomputers.com/en/Linux-Hardware/Linux-Notebooks/15-16-inch/TUXEDO-Book-XA15-Gen10.tuxedo with max specs.

**Comment by blitz** [comment]
> @blitz Thanks for your response and fix.
> 
> I tried the new changes and it behaves as follows:
> 
>     1. When I put `tuxedo-control-center` in `environment.systemPackages` in my configuration.nix it ends with the same error you said that should be sent upstream

Ah, I didn't realize that you don't use the module. Without the module, you don't get the TCC daemon and there is some dbus magic missing. So this is not expected to work.

> Edit: now it launches successfully using https://github.com/blitz/tuxedo-nixos/#usage I guess I had to kill the previous process

Yes, when you start TCC without the daemon and dbus magic it gets stuck in a weird state. I don't think there is anything we an do about it.

> Now a review of the app:

It would be interesting, if the limitations you see also happen on Ubuntu or TuxedoOS. If so, this is an upstream limitation.

**Comment by dasj19** [comment]
> It would be interesting, if the limitations you see also happen on Ubuntu or TuxedoOS. If so, this is an upstream limitation.
Hi, I figure that the prebuilt binaries for ubuntu 20.04 predates the machine I have and there is no point in trying them. So I tried compiling on ubuntu 22.04 live, but I can't figure out how to get tccd.service as mentioned in step 3. here https://github.com/tuxedocomputers/tuxedo-control-center#development-setup

Whithout the service the app crashes right after launch. So if you know of a `tccd.service` and `tccd-sleep.service` it will be handy. Otherwise I am unable to test on Ubuntu

**Comment by blitz** [comment]
> So if you know of a tccd.service and tccd-sleep.service it will be handy.

Thanks for trying!

See here for the service files: https://github.com/tuxedocomputers/tuxedo-control-center/tree/master/src/dist-data

**Comment by blitz** [comment]
> @blitz thanks for guiding me so far. I did not succeeded in having a working build from source. The resulted `tccd` binary gives segmentation fault.
> But I installed the deb packages provided by tuxedo for Ubuntu 20.04 on my Ubuntu 22.04 live usb.
> 
> The conclusion is that there is fan control support, and none of the features are unsupported. As I mentioned previously in NixOS there were certain parts of the application that printed "This feature is unsupported". In Ubuntu everything was supported.

Did you reboot after activating the module on NixOS? The module also loads a kernel module that is required for fan control etc.

**Comment by dasj19** [comment]
> Did you reboot after activating the module on NixOS? The module also loads a kernel module that is required for fan control etc.

I don't know what happened but after reboot the TCC displays a blank window. When started from terminal I get `TUXEDO Control Center is already running`. Killing the processes and trying again does not help.

journalctl says:
`Jan 27 07:55:31 tuxedo tccd[1445]: CpuWorker: Error validating/reapplying profile => Error: Could not read value from path: /sys/devices/system/cpu/cpu24/cpufreq/cpuinfo_min_freq => Error: ENOENT: no such file or directory, open '/sys/devices/system/cpu/cpu24/cpufreq/cpuinfo_min_freq => Error: ENOENT: no such file or directory, open '/sys/devices/system/cpu/cpu24/cpufreq/cpuinfo_min_freq'`

**Comment by blitz** [comment]
@dasj19 Having cpufreq sysfs files only for some CPUs is extremely weird. Do you have an unusual kernel configuration or any other special config?

That being said, given that this discussion is not necessarily regarding packaging, I would propose to continue the rootcausing as an issue in https://github.com/blitz/tuxedo-nixos/.

**Comment by NANASHI0X74** [comment]
what's the current status of this PR? anyone currently working on it? Is it only still blocked by the upstream electron issue?

**Comment by blitz** [comment]
> what's the current status of this PR? anyone currently working on it? Is it only still blocked by the upstream electron issue?
> 

I'm still interested in getting this into Nixpkgs. But upstream using ancient Electron is still an issue. Otherwise there is the issue of checking in gigantic autogenerated Nix files.

**Comment by blitz** [comment]
I'm closing this for now. If someone wants to revive it, feel free to checkout https://github.com/blitz/tuxedo-nixos/

**Pull request closed without merging**